File: src/python/Providers/GeminiProvider.py
import json
from urllib import error, request

from src.python.Providers.ProviderStrategy import ProviderStrategy

class GeminiProvider(ProviderStrategy):

    def __init__(self, provider, model, api_key, queryString):
        super().__init__(provider, model, api_key, queryString)

    def generate(self):
        payload = json.dumps({
            "contents": [
                {
                    "parts": [
                        {"text": self.queryString}
                    ]
                }
            ]
        }).encode("utf-8")

        req = request.Request(
            self.buildEndpointString(),
            data=payload,
            method="POST",
            headers={
                "Content-Type": "application/json",
                "x-goog-api-key": self.api_key or "",
            },
        )

        try:
            with request.urlopen(req, timeout=30) as response:
                return response.read().decode("utf-8")
        except error.HTTPError as exc:
            return exc.read().decode("utf-8")

    def buildEndpointString(self):
        return f"https://generativelanguage.googleapis.com/v1beta/models/{self.model}:generateContent"

File: src/java/Providers/GeminiProvider.java
package src.java.Providers;

import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.net.URI;

public class GeminiProvider extends ProviderStrategy {
    private final HttpClient httpClient;

    public GeminiProvider(String provider, String model, String api_key, String queryString) {
        super(provider, model, api_key, queryString);
        this.httpClient = HttpClient.newHttpClient();
    }

    @Override
    public String generate() {
        try {
            String json = "{\"contents\":[{\"parts\":[{\"text\":\"" + escapeJson(this.getQuery()) + "\"}]}]}";

            HttpRequest.Builder reqBuilder = HttpRequest.newBuilder()
                    .uri(URI.create(buildEndpointString()))
                    .header("Content-Type", "application/json");

            String key = getApiKey();
            reqBuilder.header("x-goog-api-key", key);

            HttpRequest request = reqBuilder
                    .POST(HttpRequest.BodyPublishers.ofString(json))
                    .build();

            HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
            return response.body();
        } catch (Exception e) {
            e.printStackTrace();
            return "{\"error\":\"" + escapeJson(e.getMessage()) + "\"}";
        }
    }

    private String buildEndpointString() {
        return "https://generativelanguage.googleapis.com/v1beta/models/" + getModel() + ":generateContent";
    }

    private static String escapeJson(String s) {
        if (s == null)
            return "";
        return s.replace("\\", "\\\\").replace("\"", "\\\"");
    }
}